### DISCUSSION TITLE
Gdb compatibility on windows

### DISCUSSION THREAD

**Discussion by aelio** [mse-org/mseide-msegui#49] [Q&A]
Hi, though overall gdb interaction has been improved, and application start chance has been raised to about 50% (but still I have to run target and then kill gdb several times just to start the debug) , that all comes at cost of inability to set or remove breakpoints when program runs, and a lot of 'read past end of file' (and process death) while stepping. 
I've tried some gdb versions under msys and mingw and they give all the same result.
Surprisingly I don't see a lot of people complying about this issue.
Does anybody know which exact versions of gdb are working best with mseide?

Update: these issues happen with x64 version running any 32 or 64 gdb. 
32bit mseide allows to set/remove breakpoint without such problems

**Comment by fredvs** [comment]
Hello @aelio

Here on Linux 64 bit with GDB 12.1 I did not find problems.
[EDIT] Not true, there are problems.

In end of readme.md there is a note:

    If there is a debugger timeout at startup enlarge the 'Project'-'Options'-'Debugger'-'Target'-'Wait before connect' value.

There are also lot of others parameters in 'Project'-'Options'-'Debugger' to tune gdb for mseide.

I have to find a Windows machine to test it (but you may also use parameters of 'Project'-'Options'-'Debugger).

**Reply by aelio** [reply]
Hi, @fredvs thanks for comment, yet I'd get away with 32bit ide. 
The error in 64 bit version, when setting/removing breakpoint is displayed as 'sigsegv in NtWaitForWorkViaWorkerFactory'

**Comment by fredvs** [comment]
Hello @aelio.

Is it possible for you to give all the steps you did to get the error _NtWaitForWorkViaWorkerFactory_ ?
I was not able to get it.

Many thanks.

Fred

**Reply by aelio** [reply]
@fredvs 

Debug something like this and try to place break at 'sleep' line. 
This reproduces the issue with x64 ide and any recent msys 64 or 32 gdb.exe as debugger

`program bugs;

{$mode objfpc}{$h-}
{$MACRO ON}
{$define cdiv := div}
{$COPERATORS ON}
{$calling register}
{$asmmode intel}
{$modeswitch nestedprocvars}

uses  sysutils;    
var    i : longint;
 begin
 for i := 0 to 99999 do
    sleep(1000);             
end.`

Build options
`-l -Mobjfpc -Sh -a -Si -CcPASCAL -Xm -v0 -Sg -Xg -g -gw`

Though I'm using fpc 331 nightly build, but I don't think it is the case

**Comment by fredvs** [comment]
Hum, after Linux 64 bit last upgrade I have also problem with MSEide + GDB: it does not init session any more.
;-(

But using _gdb_ in terminal `gdb /path/msegui-program ` debugging works.

**Comment by fredvs** [comment]
Hello.

I confirm, with GDB 12.1, MSEide ( and ideU too ) fails to debug and crash the ide at init of GDB.

Aaaaargh. ;-(

The good news is that installing a older version of GDB makes all work again:
http://mirrors.kernel.org/sourceware/gdb/releases/ 

I choose GDB 10.2 because it was working perfectly:
http://mirrors.kernel.org/sourceware/gdb/releases/gdb-10.2.tar.xz

Needed to install _textinfo_ package, then `./configure`, `make` and `sudo make install`.

Help will be needed to fix the problem with GDB > 10.2 (I dont have skill in GDB).

**Comment by fredvs** [comment]
Hello.

Compiled + installed from source  gdb 11.2 and all is ok with it, included breakpoint.

Compiled + installed from source  gdb 12.1 and all is ok with it, included breakpoint.
Strange so the Ubuntu version of gdb 12.1 has problems.

Compiled + installed from source  gdb 13.1 (trunk) and all is ok with it, included breakpoint.

This is for Linux 64 bit.

I have to find a Windows machine and test it.

**Comment by fredvs** [comment]
Re-hello.

Found a Windows 11 machine with last update.

Using fpc 3.2.2 64 bit and last MSEide 64 bit Windows.
The GDB used is the one provided by fpc.

Here I did not find problems with breakpoint and gdb. (see picture).
![mseide-gdb-win](https://user-images.githubusercontent.com/3421249/223155542-59b3c6dd-60fb-48ba-b854-c514ff05c1a9.png)

What gdb.exe version are you using? ( the one from fpc is very old)
![gdb-win](https://user-images.githubusercontent.com/3421249/223155714-ef9418a3-4978-4df5-b9fa-1a9ae39e9a6d.png)

**Comment by fredvs** [comment]
Oooops, I just see in the previous picture that MSEide 64 bit has used fpc 3.2.2. 32 bit for compilation.

And I dont have the Windows 11 machine any more to test with a program compiled with fpc 3.2.2. 64 bit and gdb 64 bit.

I will redo test asap I get a other Windows machine, sorry for the noise.

**Comment by fredvs** [comment]
Turn around done!

Tested on Windows 11: MSEide.exe 64 bit, compiling app with fpc.exe 3.2.2 64 bit and debug with gdb.exe 64 bit ( given by fpc).

All is ok, debugger works and breakpoint too, see image:
![mseide-gdb-64bit](https://user-images.githubusercontent.com/3421249/223179597-050e8abb-d4a4-45f1-84ae-30b25e6b6abb.png)

Here the debugger used:
![gdb-64bit](https://user-images.githubusercontent.com/3421249/223179735-0b1409b7-3a68-46e0-bf2f-958b92b02904.png)

So maybe there is a problem with your gdb.exe version or problems with fpc 3.3.1. trunk ( that I dont have the courage to compile-install on Windows, sorry).

Fred
